File: reference/CodeExamples/dotNet/CSharpExamples/Code/semaphores/SemaphoreExamples.cs
using System;
using Db4objects.Db4o;
using Db4objects.Db4o.CS;

namespace Db4oDoc.Code.Semaphores
{
    public class SemaphoreExamples
    {
        private const int Port = 1337;
        private const string UserAndPassword = "sa";

        public static void Main(string[] args)
        {
            using (IObjectServer server = Db4oClientServer.OpenServer("database.db4o", Port))
            {
                server.GrantAccess(UserAndPassword, UserAndPassword);
                GrabSemaphore();
                TryGrabSemaphore();
            }
        }

        private static void TryGrabSemaphore()
        {
            using (IObjectContainer container = OpenClient())
            {
                bool hasLock = container.Ext().SetSemaphore("LockName", 1000);
                if (hasLock)
                {
                    Console.WriteLine("Could get lock");
                }
                else
                {
                    Console.WriteLine("Couldn't get lock");
                }
                GrabSemaphore();
            }
        }

        private static void GrabSemaphore()
        {
            IObjectContainer container = OpenClient();

            // #example: Grab a semaphore
            // Grab the lock. Specify the name and a timeout in milliseconds
            bool hasLock = container.Ext().SetSemaphore("LockName", 1000);
            try
            {
                // you need to check the lock
                if (hasLock)
                {
                    Console.WriteLine("Could get lock");
                }
                else
                {
                    Console.WriteLine("Couldn't get lock");
                }
            }
            finally
            {
                // release the lock
                container.Ext().ReleaseSemaphore("LockName");
            }
            // #end example
        }

        private static IObjectContainer OpenClient()
        {
            return Db4oClientServer.OpenClient("localhost", Port, UserAndPassword, UserAndPassword);
        }
    }
}

File: reference/CodeExamples/dotNet/VisualBasicExamples/Code/semaphores/SemaphoreExamples.vb
Imports Db4objects.Db4o
Imports Db4objects.Db4o.CS

Namespace Db4oDoc.Code.Semaphores
    Public Class SemaphoreExamples
        Private Const Port As Integer = 1337
        Private Const UserAndPassword As String = "sa"

        Public Shared Sub Main(ByVal args As String())
            Using server As IObjectServer = Db4oClientServer.OpenServer("database.db4o", Port)
                server.GrantAccess(UserAndPassword, UserAndPassword)
                GrabSemaphore()
                TryGrabSemaphore()
            End Using
        End Sub

        Private Shared Sub TryGrabSemaphore()
            Using container As IObjectContainer = OpenClient()
                Dim hasLock As Boolean = container.Ext().SetSemaphore("LockName", 1000)
                If hasLock Then
                    Console.WriteLine("Could get lock")
                Else
                    Console.WriteLine("Couldn't get lock")
                End If
                GrabSemaphore()
            End Using
        End Sub

        Private Shared Sub GrabSemaphore()
            Dim container As IObjectContainer = OpenClient()

            ' #example: Grab a semaphore
            ' Grab the lock. Specify the name and a timeout in milliseconds
            Dim hasLock As Boolean = container.Ext().SetSemaphore("LockName", 1000)
            Try
                ' you need to check the lock
                If hasLock Then
                    Console.WriteLine("Could get lock")
                Else
                    Console.WriteLine("Couldn't get lock")
                End If
            Finally
                ' release the lock
                container.Ext().ReleaseSemaphore("LockName")
            End Try
            ' #end example
        End Sub

        Private Shared Function OpenClient() As IObjectContainer
            Return Db4oClientServer.OpenClient("localhost", Port, UserAndPassword, UserAndPassword)
        End Function
    End Class
End Namespace
